moxin-backend/src/store/remote.rs

```diff
@@ -211,154 +209,189 @@ fn download_file(
     }
 }
 
-pub fn download_file_from_remote(
-    client: &reqwest::blocking::Client,
-    model_id: &str,
-    file: &str,
-    local_path: &str,
+#[derive(Debug, Clone)]
+pub struct ModelFileDownloader {
+    client: reqwest::Client,
+    sql_conn: Arc<Mutex<rusqlite::Connection>>,
+    control_tx: tokio::sync::broadcast::Sender<DownloadControlCommand>,
     step: f64,
-    report_fn: &mut dyn FnMut(f64) -> anyhow::Result<()>,
-    must_cancel_fn: &dyn Fn() -> bool,
-) -> io::Result<DownloadResult> {
-    let url = format!(
-        "https://huggingface.co/{}/resolve/main/{}?download=true",
-        model_id, file
-    );
-
-    let content_length = get_file_content_length(client, &url)
-        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
-
-    download_file(
-        &client,
-        content_length,
-        &url,
-        local_path,
-        step,
-        report_fn,
-        must_cancel_fn,
-    )
 }
 
-pub fn download_file_loop(
-    sql_conn: Arc<Mutex<rusqlite::Connection>>,
-    rx: Arc<
-        crossbeam::channel::Receiver<(
+impl ModelFileDownloader {
+    pub fn new(
+        client: reqwest::Client,
+        sql_conn: Arc<Mutex<rusqlite::Connection>>,
+        control_tx: tokio::sync::broadcast::Sender<DownloadControlCommand>,
+        step: f64,
+    ) -> Self {
+        Self {
+            client,
+            sql_conn,
+            control_tx,
+            step,
+        }
+    }
+
+    pub async fn run_loop(
+        downloader: Self,
+        max_downloader: usize,
+        mut download_rx: tokio::sync::mpsc::UnboundedReceiver<(
             super::models::Model,
             super::download_files::DownloadedFile,
             Sender<anyhow::Result<FileDownloadResponse>>,
-            std::sync::mpsc::Receiver<DownloadControlCommand>,
         )>,
-    >,
-) {
-    let client = reqwest::blocking::Client::new();
-
-    while let Ok((model, mut file, tx, rx_control)) = rx.recv() {
-        let file_id = file.id.clone();
-
-        let mut send_progress = |progress| {
-            match tx.send(Ok(FileDownloadResponse::Progress(
-                file_id.as_ref().clone(),
-                progress as f32,
-            ))) {
-                Ok(_) => {}
-                Err(_) => return Err(anyhow::anyhow!("Download stopped by the receiver")),
-            };
+    ) {
+        let (token_tx, mut token_rx) = tokio::sync::mpsc::channel::<()>(max_downloader);
+        for _ in 0..max_downloader {
+            let _ = token_tx.send(());
+        }
 
-            // Update our local database
-            {
-                let conn = sql_conn.lock().unwrap();
-                let pending_download = PendingDownloads {
-                    file_id: file_id.clone(),
-                    progress: progress,
-                    status: PendingDownloadsStatus::Downloading,
```

Comment:
This code is executed frequently during downloads, and it causes a large number of writes to the database and frequently occupies locks. So, I optimized this part of the logic. I store the total length of the file in the database, and when I need to get the download progress, I calculate it based on the downloaded portion and the total length.

Reply:
Ok, I kind of have the same concern about frequent database access, although here is a situation where the database is only used by one client (unlike web servers database).

One import requirement to keep working is the ability to get this progress value (in percentage) when the whole app restarts. The user should see the paused downloaded with their corresponding progress. Agreed that it can be done by comparing the real file size and the stored total 👍